ansible/roles/docker/tasks/main.yml

```diff
@@ -38,29 +46,69 @@
   when: ansible_os_family == "RedHat"
 
 - name: check docker-tcp socket state
-  shell: systemctl status docker-tcp.socket | grep "Active:"
+  shell: systemctl status docker-tcp.socket | grep 'Active.*active' -o
   register: docker_tcp_socket_state
 
+- include: create_docker_device.yml
+  when: docker_device != ""
+
 - name: copy systemd units for docker tcp socket settings
   template: src=docker-tcp.j2 dest=/etc/systemd/system/docker-tcp.socket
   register: docker_tcp_socket
 
   # tcp socket service requires docker service to be started after it
-- name: start docker tcp socket service
-  shell: sudo systemctl stop docker && sudo systemctl start docker-tcp.socket && sudo systemctl start docker
-  when: (docker_tcp_socket | changed) or not docker_tcp_socket_state.stdout | match("Active.* active")
+- name: reload and stop docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: stop docker
+  service:
+    name: docker
+    state: stopped
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: start docker-tcp service
+  service:
+    name: docker-tcp.socket
+    state: started
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: check docker service state
+  shell: sudo systemctl status docker | grep 'Active.*active' -o
+  ignore_errors: true
+  register: docker_service_state
+  tags:
+    - prebake-for-dev
 
 - name: remove the docker key file, if any. It shall be regenerated by docker on restart
   file: name=/etc/docker/key.json  state=absent
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
 
 # XXX: service module doesn't do daemon-reload yet, so need to use shell module here
 # https://github.com/ansible/ansible-modules-core/issues/191
-- name: restart docker
+- name: reload docker
   #service: name=docker state=restarted
-  shell: sudo systemctl daemon-reload && sudo systemctl restart docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
   tags:
     - prebake-for-dev
 
+# XXX: this needs to happen twice after setting up the thin provisioner because
+#      of some docker bug I've not investigated.
+- name: restart docker (first time)
+  service:
+    name: docker
+    state: restarted
+  when: 
```

Comment:
> No, if the var doesn't go away like you keep suggesting, this never happens.

hmm, yes, you are right the var should have been defined. I looked at the logs again (`make provision` case) and this is what I saw which I assumed was coming from this variable being undefined but I was mistaken. I think there is still something that might need to be fixed in the condition but I might be missing something as you didn't see it.

```
TASK [docker : restart docker (first time)] ************************************
fatal: [mon0]: FAILED! => {"failed": true, "msg": "The conditional check 'thin_provisioned and thin_provisioned.rc == 0' failed. The error was: error while evaluating conditional (thin_provisioned and thin_provisioned.rc == 0): 'dict object' has no attribute 'rc'\n\nThe error appears to have been in '/home/ladmin/goroot/src/github.com/contiv/volplugin/ansible/roles/docker/tasks/main.yml': line 102, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n#      of some docker bug I've not investigated.\n- name: restart docker (first time)\n  ^ here\n"}
...ignoring
TASK [docker : ensure docker is started] ***************************************
changed: [mon0]
```